narwhal/primary/src/state_handler.rs

```diff
@@ -185,16 +250,184 @@ impl RandomnessState {
                 &mut rand::thread_rng(),
             ) {
                 Ok(output) => {
+                    if let Err(e) = self.vss_key_output.set(output.vss_pk.clone()) {
+                        error!("random beacon: unable to write VSS key to output: {e:?}")
+                    }
                     self.dkg_output = Some(output);
                     info!(
                         "random beacon: DKG complete with Output {:?}",
                         self.dkg_output
                     );
                 }
                 Err(fastcrypto::error::FastCryptoError::NotEnoughInputs) => (), // wait for more input
-                Err(e) => error!("Error while processing randomness DKG confirmations: {e:?}"),
+                Err(e) => error!("random beacon: error while processing DKG Confirmations: {e:?}"),
+            }
+            // Begin randomness generation.
+            if self.dkg_output.is_some() {
+                self.send_partial_signatures().await;
+            }
+        }
+    }
+
+    async fn update_narwhal_round(&mut self, round: Round) {
+        self.narhwal_round = round;
+        let next_leader_narwhal_round = round + (round % 2);
+        // Re-send partial signatures to new leader, in case the last one failed.
+        if self.dkg_output.is_some() && (next_leader_narwhal_round > self.last_narwhal_round_sent) {
+            self.send_partial_signatures().await;
+        }
+    }
+
+    async fn update_randomness_round(&mut self, round: u64) {
+        if round <= self.randomness_round {
+            // Don't go backwards.
+            return;
+        }
+        self.randomness_round = round;
+        self.partial_sigs
+            .retain(|&(round, _), _| round >= self.randomness_round);
+        self.send_partial_signatures().await;
+    }
+
+    async fn send_partial_signatures(&mut self) {
+        let dkg_output = match &self.dkg_output {
+            Some(dkg_output) => dkg_output,
+            None => {
+                error!("random beacon: called send_partial_signatures before DKG completed");
+                return;
+            }
+        };
+        let shares = match &dkg_output.shares {
+            Some(shares) => shares,
+            None => return, // can't participate in randomness generation without shares
+        };
+        debug!(
+            "random beacon: sending partial signatures for round {}",
+            self.randomness_round
+        );
+        let sigs = ThresholdBls12381MinSig::partial_sign_batch(
```

Comment:
This part is expensive - should we also store those in the state instead of recomputing them more than once in case the next leader is slow?